cryptol-saw-core/src/Verifier/SAW/TypedTerm.hs

```diff
@@ -67,22 +68,35 @@ mkTypedTerm sc trm = do
         Just (Right t) -> TypedTermSchema (C.tMono t)
   return (TypedTerm ttt trm)
 
--- | Apply a function-typed 'TypedTerm' to an argument. This operation
--- fails if the first 'TypedTerm' does not have a monomorphic function
--- type.
+-- | Apply a function-typed 'TypedTerm' to an argument.
+--   This operation fails if the type of the argument does
+--   not match the function.
 applyTypedTerm :: SharedContext -> TypedTerm -> TypedTerm -> IO TypedTerm
-applyTypedTerm sc (TypedTerm tp t1) (TypedTerm _ t2)
-  | Just (_,cty') <- C.tIsFun =<< ttIsMono tp
-  = TypedTerm (TypedTermSchema (C.tMono cty')) <$> scApply sc t1 t2
-
--- TODO? extend this to allow explicit application of types?
-applyTypedTerm _ _ _ = fail "applyTypedTerm: not a (monomorphic) function type"
+applyTypedTerm sc (TypedTerm _tp t1) (TypedTerm _ t2) =
+  do trm <- scApply sc t1 t2
+     ty <- scTypeCheckError sc trm
+     ct <- scCryptolType sc ty
```

Comment:
IIRC `scCryptolType` can be a bit weird sometimes due to the lack of an injective mapping from Cryptol types to SAWCore types. I think this might present some unintuitive behavior when (e.g.) the `Term` applied is a Cryptol function returning a tuple or record. I don't think this is a blocker, but it might be worth mentioning in the docs for `term_apply` or even just as a comment here.